harness/determined/keras/_tf_keras_trial.py

```diff
@@ -356,36 +319,158 @@ def from_native(
+
+        self._train_input_manager, self._validation_input_manager = keras._init_input_managers(
+            context=self.context, train_config=train_config
         )
 
-        if hvd_config.use and version.parse("2.0.0") <= version.parse(
-            tf.__version__
-        ) < version.parse("2.2.0"):
-            logging.info(
-                "Calling `model.compile(...)` with `experimental_run_tf_function=False` to ensure "
-                "TensorFlow calls `optimizer.get_gradients()` to compute gradients."
-            )
+        self._configure_callbacks(train_config.callbacks, train_config.verbose)
 
-            context.model.compile(
-                *compile_args.args, **compile_args.kwargs, experimental_run_tf_function=False
-            )
-        else:
-            context.model.compile(*compile_args.args, **compile_args.kwargs)
+        # If a load path is provided, load weights and restore the data location.
+        self._load()
+
+        self.train_response_func = None  # type: Optional[workload.ResponseFunc]
+        self.train_workload_metrics = []  # type: List[Dict[str, Any]]
+        self.train_workload_batches = 0
+        self.train_workload_inputs = 0
+        self.train_workload_len = 0
+        self.test_inputs = 0
+
+    def _configure_callbacks(self, user_callbacks: Optional[List], verbose: int) -> None:
+        """
+        If we pass a callbacks parameter to model.fit() or model.evaluate() which is a
+        pre-constructed CallbackList, Keras will not alter it.  We can use this property to
+        configure the exact callback order that we want in our system.
+
+        The implementation is based closely on from the real
+        tf.keras.callbacks.configure_callbacks(), with the following differences:
+
+          - We always assume we have the original Callbacks list.
+          - We prepend and append additional Determined and Horovod callbacks
+          - We create a det.keras.CallbackList instead of the normal tf.keras one.
+        """
+
+        callbacks = user_callbacks or []
+        check.is_instance(
+            callbacks,
+            list,
+            "the callbacks parameter of model.fit() or model.eval() must be a list of Callbacks",
+        )
+
+        if self.env.experiment_config.get_records_per_epoch() is None:
+            for cb in callbacks:
+                if util.is_overridden(cb.on_epoch_end, tf.keras.callbacks.Callback):
+                    if not isinstance(cb, keras.callbacks.Callback):
+                        # New callbacks must obey the rules.
+                        raise AssertionError(
+                            "it is unsupported to use a Callback that defines on_epoch_end "
+                            f"({type(cb).__name__}) without setting the records_per_epoch value "
+                            "in the experiment config"
+                        )
+                    else:
+                        # Pre-existing callbacks only get a warning.
```

Comment:
This logic is new since you last reviewed.

Reply:
I'm honestly not sure if that's a good idea or not.  I feel like it does increase the initial porting effort